### ISSUE TITLE
Allow null input values in Query.t_variables

### ISSUE THREAD

**Issue by osener** [reasonml-community/graphql-ppx#26]
Hi, first of all thanks for continuing maintenance of this useful ppx!

I'd like to migrate to `graphql_ppx_re` for BuckleScript 6 support, but my app breaks because of the addition of `filter_out_null_values` to outputted code. Apart from introducing polymorphic comparison, my GraphQL server implemented using `ocaml-graphql-server` don't seem to like missing variables.

Here's a diff comparing graphql_ppx and graphql_ppx_re outputs compiled using BuckleScript:
```diff
   return {
           query: ppx_printed_query,
           variables: Js_dict.fromArray(/* array */[/* tuple */[
-                  "after",
-                  after !== undefined ? Caml_option.valFromOption(after) : null
-                ]]),
+                    "after",
+                    after !== undefined ? Caml_option.valFromOption(after) : null
+                  ]].filter((function (param) {
+                      return Caml_obj.caml_notequal(param[1], null);
+                    }))),
           parse: parse
         };
```

When I pass `None` for this `"after"`  variable which stands for an optional GraphQL Cursor, I get `Missing variable 'after'` back from the server.

Does this change implement something that's spec-compliant and highlights a problem in `ocaml-graphql-server`, or is there another reason for it?

**Comment by baransu** [comment]
Variables are represented as JSON and JSON has no concept of `undefined` variables. That's why JS object  `{ a: undefined, b: "value" }` will result in  `{"b": "value"}` JSON.

According to GraphQL spec: https://graphql.github.io/graphql-spec/June2018/#sec-Input-Objects
> If no value is provided for a defined input object field and that field definition provides a default value, the default value should be used. If no default value is provided and the input object field’s type is non‐null, an error should be thrown. Otherwise, if the field is not required, then no entry is added to the coerced unordered map.

So looking at GraphQL spec our implementation is correct. From my experience, it works correctly in: 
- https://github.com/absinthe-graphql/absinthe
- https://github.com/sangria-graphql/sangria

**Comment by jfrolich** [comment]
It looks however if it's filtering out null, null is a valid value according to the spec, so no value and null have a different semantic meaning. The spec link you posted reflects that as well. Absinthe also correctly deals with null/missing in my experience.

**Comment by baransu** [comment]
It's problematic in Reason/OCaml semantics because we don't have third value to represent `None | Some('a)`. There was an initial idea to represent it as `option(option('a))` but IMO it's not good. 

So based on spec `null` means the value was provided but was `null`, and no value means it was not provided at all. But is there a difference from the server-side point of view in Absinthe or in both cases you receive input value as `nil`?

I'm open to redesigning how we handle not provided value vs explicitly provided `None` but an original issue, when the value is not provided is not spec compliant from `ocaml-graphql-server` point of view as it should handle not provided nullable fields.

We slightly stretch spec by missing semantics a little bit but I'm not sure how we should handle such cases.

cc @anmonteiro @Schmavery

**Comment by jfrolich** [comment]
> So based on spec `null` means the value was provided but was `null`, and no value means it was not provided at all. But is there a difference from the server-side point of view in Absinthe or in both cases you receive input value as `nil`?

In absinthe it is a way to set a value to `nil` or empty. If the field is not provided it means that the value is not being changed. If `nil` is provided, it means make the field empty. For instance in a multi-select.

Can we not have the same functionality with the make function (optional fields not provided yields an object without the key, and when None is explicitly provided it yields a field with a `null` value).

**Comment by baransu** [comment]
Using `make` function would be not that different but input types would force us to use `{"field": Some(Some(value))}` which is not that pretty. But overall I think `option(option(t))` would be the best option.

**TL;DR**
None -> field is omitted
Some(None) -> null
Some(Some(value)) -> value

Example: 
```graphql
type Address {
  street: String!
  houseNumber: String
}

mutation Update($firstName: String, $address: Address!) {
  update(firstName: $firstName, address: $address)
}
```

```reason
Mutation.make(~address={"street": "Private Street", "houseNumber": None})##variables;
// {"address": {"street": "Private Street"}}

Mutation.make(~firstName=None, ~address={"street": "Private Street", "houseNumber": None})##variables;
// {"firstName": null, "address": {"street": "Private Street"}}

Mutation.make(~address={"street": "Private Street", "houseNumber": Some(None)})##variables;
// {"address": {"street": "Private Street", "houseNumber": null }}
```

**Comment by baransu** [comment]
> In absinthe it is a way to set a value to nil or empty. If the field is not provided it means that the value is not being changed. If nil is provided, it means make the field empty. For instance in a multi-select.

I mean from when receiving mutation arguments in field resolver. How both cases are represented?

**Comment by jfrolich** [comment]
Yeah I actually think `Some(None)` makes sense semantically when None can be provided explicitly. I am still a reasonml beginner, but I think with [@bs.obj] it can generate an object with missing fields I believe instead of `Some(None)`.

**Comment by jfrolich** [comment]
> > In absinthe it is a way to set a value to nil or empty. If the field is not provided it means that the value is not being changed. If nil is provided, it means make the field empty. For instance in a multi-select.
> 
> I mean from when receiving mutation arguments in field resolver. How both cases are represented?

Ah ok :) It's basically the same, a map with missing values or a map with a `nil` value.

**Comment by baransu** [comment]
Thanks, that makes sense 🙂

**Comment by osener** [comment]
I think ocaml-graphql-server fails early because a declared variable in the query does not exist in the `variables` dictionary. It might be an easy change, but I haven't look at the source of the problem yet to see if the type information (nullability) is available at the point where this error is thrown.

@baransu the old version already does the right thing regarding `None -> null` translation. What's the advantage of dropping the null values altogether?

**Comment by baransu** [comment]
@osener I think Sangria + not sure which JSON decoder was unable to handle null values in nested types and I fixed it in as a workaround which ended up in this fork. Old version and current version are both valid but not ideal and not 100% spec-compliant.

**Comment by osener** [comment]
Thanks for looking into this. In that case I'll need to stick to graphql_ppx for the time being. 

Feel free to close the issue if you don't plan to revert this behavior or allow opting-out.

**Comment by baransu** [comment]
@osener If the declared variable in the query does not exist in variables dictionary there is pending PR for such case: https://github.com/baransu/graphql_ppx_re/pulls

As regard nullability, I plan to move to `option(option(t))` as it will allow us to be spec-compliant. I'll ping you to let you know 🙂

**Comment by osener** [comment]
I know you know this, but just to summarize, this is the opposite of #25;

I'm constructing a query with `Query.makeWithVariables({"after": None})`, and seeing keys with `None` (`null` in the end) as values get dropped from the serialized variables dictionary, which my server implementation does not like.

https://github.com/andreas/ocaml-graphql-server/blob/2dcd97d0248ec01a798390c60dfa7721c0d61729/graphql/src/graphql_schema.ml#L289

**Comment by jfrolich** [comment]
Thanks for the great work. I'll hope to add some PR's to this codebase too (many ideas), once I understand it 😱

**Comment by osener** [comment]
Hi @baransu. I'd like to help get this issue resolved, but from our discussion above I assume reverting this logic back to the `graphql_ppx` version would not work for you right?

**Comment by baransu** [comment]
So both options - current impl and previous impl are valid and invalid at the same time. Previous impl assumed all not specified values are `null`, currently assumes they are `undefined`.
I think the best solution should cover both cases thus `option(option('a))` seems to fit these criteria.

**Comment by peterpme** [comment]
Hi there,

I'm using apollo-server and the GraphQL playground. The GraphQL playground works great, but the null stripping forces us to create a new mutation that removes the keys which is a little annoying.

Just wanted to resurface this to see if there's been any progress. Thanks for your hard work!

**Comment by nirvdrum** [comment]
I think in an ideal world, a blended approach would work out. If the value isn't supplied in `makeVariables`, it should be `undefined`, if `None` then `null`, or if `Some(value)` then `value`. I realize that doesn't really work out so well with the type system though.

While I can see how `option(option('a))` would make sense, I suspect it'll make code a bit harder to read. I think a variant type with the three cases spelled out would make client code easier to read. There's probably not much appetite here to raise the minimum supported OCaml version, but `@unboxed` can provide nice solutions in cases like this.

**Comment by jfrolich** [comment]
Yeah. It's not too pretty. The first option is not possible with Reason, because not supplied is the same as `None`. We can indeed have a switch in the query to explicitly enable the ability to supply `null`. Variants would work, but perhaps it's better to be idiomatic Bucklescript by passing a `Js.Nullable.t`?

**Comment by nirvdrum** [comment]
I veered away from `Js.Nullable.t` just because GraphQL's type system is separate from JavaScript's. Maybe that's a distinction that doesn't really matter.

**Comment by jfrolich** [comment]
> I veered away from `Js.Nullable.t` just because GraphQL's type system is separate from JavaScript's. Maybe that's a distinction that doesn't really matter.

Yes but we need three variants. `option(option('a))` is the alternative, we can also potentially do a custom   variant like `None` / `Null` / `Some('a)` but that might have some other drawback.

**Comment by jfrolich** [comment]
So we now expose the raw argument type, which needs to be a `Js.Nullable.t`, so if you use this directly, you are now able to pass null values.

I think for the unserialized type it would be best to have an option to encode the nullable arguments to a variant with three options instead of two (as in the `option` type). So leaving this open.

**Comment by jfrolich** [comment]
I think current solution is good enough.
